Safety/Src/safety_controller.cpp

```diff
@@ -1,44 +1,140 @@
 #include "safety_controller.hpp"
 #include "PWM.hpp"
 #include "stdlib.h"
-//#include "Interchip.h"
 #include "interchip_S.hpp"
+#include "RSSI.hpp"
+
+#include "gpio.h"
+
+/***********************************************************************************************************************
+ * Definitions
+ **********************************************************************************************************************/
+
+constexpr int MANUAL_OVERRIDE_CHANNEL = 4;
+// temporary grabber_rx and grabber_pwm channels. To be changed later.
+constexpr int grabber_rx_channel = 0;
+constexpr int grabber_pwm_channel = 9;
+
+
+        // pwm order: left tail, right tail, aileron, nose wheel, throttle, gimbal x, gimbal y, reverse thrust.
+#define PWM_CHANNEL_LEFT_TAIL 0
+#define PWM_CHANNEL_RIGHT_TAIL 1
+#define PWM_CHANNEL_AILERON 2
+#define PWM_CHANNEL_NOSE_WHEEL 3
+#define PWM_CHANNEL_THROTTLE 4
+#define PWM_CHANNEL_GIMBAL_YAW 5
+#define PWM_CHANNEL_GRABBER 6
+
+/***********************************************************************************************************************
+ * Prototypes
+ **********************************************************************************************************************/
 
 static void setPWMChannel(PWMChannel &pwm, int channel, int percentage);
 static int getPPM(PPMChannel &ppm, int channel);
 
+static bool AutoPilotEngaged(PPMChannel &ppm);
 
-
-//static Interchip_AtoS_Packet *dataRX;
-//static Interchip_StoA_Packet *dataTX;
+/***********************************************************************************************************************
+ * Code
+ **********************************************************************************************************************/
 
 void safety_controller_init()
 {
-
-    //dataTX = (Interchip_StoA_Packet*) malloc(sizeof(Interchip_StoA_Packet));
-    //dataRX = (Interchip_AtoS_Packet*) malloc(sizeof(Interchip_AtoS_Packet));
-    //Interchip_Init(dataTX, dataRX);
 }
 
 void safety_run(PWMChannel &pwm, PPMChannel &ppm)
 {
-    if(!isSafetyManual())
+
+    if(CommsFailed()) {
+        setPWMChannel(pwm, PWM_CHANNEL_LEFT_TAIL, 0);
+        setPWMChannel(pwm, PWM_CHANNEL_RIGHT_TAIL, 100);
+        setPWMChannel(pwm, PWM_CHANNEL_AILERON, 100);
+        setPWMChannel(pwm, PWM_CHANNEL_THROTTLE, 0);
+        setPWMChannel(pwm, PWM_CHANNEL_GIMBAL_YAW, 50);
+        setPWMChannel(pwm, PWM_CHANNEL_NOSE_WHEEL, 0);
+
+        return;
+    }
+
+    volatile int16_t *AutoPilotPwmChannel = getPWM();
```

Comment:
Why is this volatile?

Reply:
@AadityaChaudhary

Reply:
Correct me if im wrong, but getPWM is marked volatile because it should be returning an array of PWM values from the Interchip packet (which should be volatile). I think thats why this variable is marked volatile.